c2rust-analyze/src/rewrite/expr/unlower.rs

```diff
@@ -0,0 +1,379 @@
+//! Builds the *unlowering map*, which maps each piece of the MIR to the HIR `Expr` that was
+//! lowered to produce it.
+//!
+//! For example:
+//!
+//! ```Rust
+//! fn f(a: i32) -> i32 {
+//!     a + 1
+//! }
+//!
+//! fn g(x: i32, y: i32) -> i32 {
+//!     x + f(y)
+//! }
+//! ```
+//!
+//! For `f`, the unlowering map annotates the MIR as follows:
+//!
+//! ```text
+//! block bb0:
+//!   bb0[0]: StorageLive(_2)
+//!   bb0[1]: _2 = _1
+//!     []: StoreIntoLocal, `a`
+//!     [Rvalue]: Expr, `a`
+//!   bb0[2]: _0 = Add(move _2, const 1_i32)
+//!     []: StoreIntoLocal, `a + 1`
+//!     [Rvalue]: Expr, `a + 1`
+//!   bb0[3]: StorageDead(_2)
+//!   bb0[4]: Terminator { source_info: ..., kind: return }
+//! ```
```

Comment:
The format used here isn't very clear at all.  If you could add a bit of an explainer, that'd be great and make it a lot easier to understand, or at least intra-doc link to the code that prints this format.

Reply:
I've slightly expanded the paragraph below in 85959eeaf802e208747da14b34fe5007a98e6018 - let me know if that helps.

Reply:
Yeah, that's a lot clearer now!

By the way, I generally try to have line breaks at semantic breaks in the sentence vs. hard wrapping at a certain length.  That way diffs are much cleaner and don't spill over to other unchanged lines.  Also, it's IMO easier to read.

Reply:
>I generally try to have line breaks at semantic breaks in the sentence vs. hard wrapping at a certain length. That way diffs are much cleaner and don't spill over to other unchanged lines.

I used to do the same, but it turns out `git diff --word-diff` does a pretty good job of unwrapping the diff:

```
/// {+Each statement is shown here with its MIR `Location`, such as `bb0[0]`; within each statement,+}
{+/// annotations are associated with particular `SubLoc` paths, such as `[]` or `[Rvalue]`.+}  The
{+///+} statement `_2 = _1` is associated with the expression `a`; the statement[-///-] as a whole {+(`bb0[1]`+}
{+/// `[]`)+} is storing the result of evaluating `a` into a MIR local, and the[-///-] statement's rvalue
{+/// (`bb0[1]` `[Rvalue]`)+} `_1` represents the expression `a` itself.  Similarly, `_0 =[-///-] Add(move _2,
{+///+} const 1)` stores the result of `a + 1` into a local.  If needed,[-///-] we could extend the `unlower`
{+///+} pass to also record that `move _2` (a.k.a. `bb0[2]`[-///-] `[Rvalue, RvalueOperand(0)]`) is lowered
{+///+} from the `Expr` `a`.
```

(Looks nicer in color - try `git show 85959ee --word-diff`)

Reply:
I use that, too, but I don't know how to get GitHub to use it (that'd be amazing!).